### PR TITLE
feat(compiler)!: Remove Exclusive/Inclusive Ranges, provide as record via compiler

### LINKED ISSUE #1404: Range enum should be provided by compiler
We've found the need to use the `Range` enum in the number library. It seems really heavy-handed to include the Range module for the data type.

We should move the `Range` enum to Pervasives and re-export it from the Range module (for backwards compatibility).

Eventually, `Range` could be a data type implemented inside the compiler and maybe even receive syntax.

### PR DESCRIPTION
**Pull request by phated**
Closes #1404 

This adds the Range enum as a language-supplied type, which is a prerequisite to having language syntax for ranges.

The new Range is a `record Range<a> { rangeStart: a, rangeEnd: a }`. It is now up to APIs to decide if the range is interpreted as Inclusive or Exclusive, and is no longer limited to only Numbers, which also allow APIs to decide how to interpret the types.

### CODE REVIEW

**Comment by alex-snezhko** [comment]
I think if `Exclusive` ranges were to be kept then special syntax should be added for them, otherwise I think it'd be better to remove them altogether as it's awkward to have syntax for one way but not the other (also doing `end - 1` for exclusive ranges isn't too big of a deal at the end of the day).

Also, thoughts on adding unbounded ranges like `3..`? I find them pretty useful in other languages that have them for slicing and whatnot (this would also allow the various `slice` functions to take a `Range` rather than a second optional argument as discussed in [#1566](https://github.com/grain-lang/grain/issues/1566#issuecomment-1369090465) which seems nicer imo). `Range.map` could then be replaced with a generator function instead of returning a list to accommodate for not being able to create an infinite list (or an exception could just be thrown in this case :shrug: )

**Comment by phated** [comment]
I've updated this PR to turn `Range` into `record Range<a> { range_start: a, range_end: a }` as we discussed on the community call. It is now up to APIs to decide if the range is interpreted as Inclusive or Exclusive, and is no longer limited to only Numbers, which also allow APIs to decide how to interpret the types.

**Comment by alex-snezhko** [comment]
Should the casing of the record fields be changed to camelcase to match the rest of the language?

**Comment by phated** [comment]
Probably, but I hate it. 🤣

**Comment by alex-snezhko** [comment]
What about just `start` and `end`? Since `Range` will become a builtin type then maybe the `range` prefix can just be omitted by the same logic that `Some` is not called `OptionSome`, `Ok` is not called `ResultOk`, etc?

**Comment by phated** [comment]
We discussed `start` and `end` but it's not a good idea to use commonly used words in records provided by the compiler because they will result in ambiguous record warnings when people write custom records with those common words.

**Comment by ospencer** [comment]
Thoughts on `Inclusive` and `Exclusive` submodules (or just an inclusive submodule)? I feel like it'd be pretty nice to be able to do `Range.Inclusive.inRange()` in those times I want an inclusive range.

**Comment by phated** [comment]
>  just an inclusive submodule

I like the idea. I suppose I need to go back and get that implementation 😛

File: compiler/src/typed/builtin_types.re [hunk 1]
```diff
@@ -95,6 +99,8 @@ and type_option = var =>
 and type_result = (ok, err) =>
   newgenty(TTyConstr(path_result, [ok, err], ref(TMemNil)))
 and type_list = var => newgenty(TTyConstr(path_list, [var], ref(TMemNil)))
+and type_range = var =>
+  newgenty(TTyRecord([("rangeStart", var), ("rangeEnd", var)]))
```
Comment (ospencer) [inline compiler/src/typed/builtin_types.re:103]: It's insane for me to suggest this right?
```suggestion
  newgenty(TTyRecord([(Ident.name(ident_range_start), var), (Ident.name(ident_range_end), var)]))
```

File: stdlib/range.gr [hunk 2]
```diff
@@ -126,3 +111,106 @@ provide let map = (fn, range) => {
...
+   * @param range: The range to check within
+   * @returns Whether or not the value is within range
+   *
+   * @example Range.Inclusive.inRange(1, { rangeStart: 0, rangeEnd: 1 }) == true
+   * @example Range.Inclusive.inRange(10, { rangeStart: 0, rangeEnd: 2 }) == false
+   *
+   * @since v0.6.0
+   * @history v0.3.0: Root APIs originally handled Inclusive & Exclusive variants
+   */
+
```
Comment (ospencer) [inline stdlib/range.gr:129]: Can the docs touch the implementations or is this a formatter bug?

**Review by ospencer — COMMENTED** [review]
This looks awesome!

File: stdlib/range.gr (hunk 2 above)
Comment (phated) [reply stdlib/range.gr:129]: Formatter bug. We're not blocking on it though. Hopefully it'll be fixed soon.

File: compiler/src/typed/builtin_types.re (hunk 1 above)
Comment (phated) [reply compiler/src/typed/builtin_types.re:103]: I don't know the difference. I thought this type took a string?

Comment (phated) [reply compiler/src/typed/builtin_types.re:103]: Oh, I read that wrong, it's pulling the string name from the ident

**Review by ospencer — APPROVED** [review]
Excellent!

**Pull request merged**